rfcs/20200205-standalone-keras-repository.md

```diff
@@ -0,0 +1,351 @@
+|Can't easily delete keras project, which already have tons of stars and incoming reference links. Continued existence of external Keras code will create confusion ("why is there tensorflow/keras AND keras-team/keras?")|Issue/PR under the same org can be transferred easily, but not cross the different org. See here|
+
+### Source of Truth
+
+TensorFlow uses a Google-internal code repository as its source of truth. Every PR
+submitted though GitHub is converted to a Google-internal change first,
+submitted through the internal system, and then copied to GitHub as commits.
+At the same time, PR is marked as merged with the corresponding commit hash.
+
+Likewise, issue tracking and code review takes place through Google-internal tools.
+
+For Keras, since we are trying to promote community engagement, we hope to use 
+GitHub as source of truth. This will have the following implications:
+
+* We expect the majority of the code development/contribution from GitHub
+and the dev tools / tests / scripts should focus on the GitHub development use
+case. See below for more details.
+* Keras CI/presubmit build for the GitHub repo should target the `tf-nightly` pip
+package as dependency. This means any change to TF will take at most 24
+hours to be reflected on the Keras side.
+* The Keras code will be mirrored to a Google-internal code repository via Google-internal 
+tools within a very short time window after each change.
+The Google-internal CI tests will run on HEAD for both Keras and TF code.
+* The CI build for the repository on GitHub might break when it sees a
+new version of `tf-nightly`, if  certain behavior has been changed and wasn't
+caught by unit tests. We have  observed a few similar cases with
+[tf/addons](https://github.com/tensorflow/addons).
+We hope this can be reduced by stronger
+unit test coverage by Google internel systems, when both TF and Keras code are 
+tested at HEAD.
+* pip package management. Keras will now follow the `tf-estimator` approach. 
+"pip install tensorflow" should also install Keras (from PyPI) as well.
+There are more details for the pip package in the
+[Improved pip package structure](https://github.com/tensorflow/community/pull/182) RFC.
+
+### Dependency Cleanup
+
+As the high-level API of TensorFlow, Keras should have a direct dependency on
+TF low-level APIs, but not the other way around. Unfortunately, there is some existing reverse 
+logic in the TF code that relies on Keras, which we should update/remove 
+when we split the repository.
+
+The current usage of Keras from TensorFlow are:
+* Unit tests, which should be converted to integration tests, or port the tests
+to Keras repository.
+* `feature_column`.
+* Legacy `tf.layers` in v1 API.
+* legacy RNN cells.
+* TPU support code for `optimizer_v2`.
+* SavedModel.
+* TF Lite.
+
+All Keras imports in integration tests can be changed to use dynamic import like below:
```

Comment:
We should be pretty careful about how we structure the integration tests & what we make them depend on to avoid this where possible.

Reply:
I have strong reservations about this type of cross-package circular dependency.

Reply:
Indeed, I would also like to cleanup this dependency before we split the repository.